feat(forms): add ValidatorDirective

vsavkin opened the pull request.

rkirov commented on modules/angular2/src/core/forms/directives/ng_model.ts:
```diff
@@ -75,7 +76,7 @@ export class NgModel extends NgControl implements OnChanges {
   get path(): string[] { return []; }
 
-  get validator(): Function { return Validators.compose(this.validators); }
+  get validator(): Function { return this._validator; }
```
same code repeated 3 times, have you thought about extracting into a base class?

vsavkin replied:
This would require storing state in a superclass, which I'm trying to avoid. It makes thinking about the code more difficult.

rkirov commented on modules/angular2/src/core/forms/directives/normalize_validator.ts:
```diff
@@ -0,0 +1,10 @@
+import {Control} from '../model';
```
unused

vsavkin replied:
Fixed

rkirov commented on modules/angular2/src/core/forms/directives/normalize_validator.ts:
```diff
@@ -0,0 +1,10 @@
+import {ValidatorDirective} from './validators';
+
+export function normalizeValidator(validator: Function | ValidatorDirective): Function {
+  if ((<ValidatorDirective>validator).validate !== undefined) {
```
why not trust TS and do 

```
if (validator instanceOf Function) return validator;
else (c) => (<ValidatorDirective>validator).validate(c)
```
the assumption being that if not function we do have ValidatorDirective type.

vsavkin replied:
class ValidatorDirective {}
var v = new ValidatorDirective();
v instanceof Function will be true

rkirov replied:
http://www.typescriptlang.org/Playground#src=class%20ValidatorDirective%20%7B%7D%0Avar%20v%20%3D%20new%20ValidatorDirective()%3B%0Aconsole.log(v%20instanceof%20Function)

rkirov commented on modules/angular2/src/core/forms/directives/validators.ts:
```diff
@@ -2,8 +2,16 @@ import {forwardRef, Provider, OpaqueToken} from 'angular2/src/core/di';
 
+
+/**
+ * ADD DOCS
```
TODO: add docs

rkirov commented on modules/angular2/src/core/forms/directives/validators.ts:
```diff
@@ -2,8 +2,16 @@ import {forwardRef, Provider, OpaqueToken} from 'angular2/src/core/di';
+/**
+ * ADD DOCS
+ */
+export interface ValidatorDirective { validate(c: modelModule.Control): {[key: string]: any}; }
```
Directive is a bit confusing as the interface can be implemented by non directives. Your call.

rkirov commented on modules/angular2/test/core/forms/directives_spec.ts:
```diff
@@ -48,6 +49,10 @@ class DummyControlValueAccessor implements ControlValueAccessor {
 }
 
+class CustomValidatorDirective implements ValidatorDirective {
+  validate(c: Control): {} { return {"custom": true}; }
```
why `{}` vs `{[key: string]: any}` as above

rkirov replied:
I fact i have as sneaking suspicion that all TS objects have `[key: string]: any` implicitly.
The following typechecks:

```
class F{}

var f1: {[key: string]: any} = '';
var f2: {[key: string]: any} = 3;
var f3: {[key: string]: any} = [];
var f4: {[key: string]: any} = {};
var f5: {[key: string]: any} = () => null;
var f6: {[key: string]: any} = new F();
```

vsavkin replied:
Just a bad copy-paste.

>> The following typechecks
Interesting.

rkirov commented:
Looks good, but I have no idea why are you doing this. Please add an appropriate commit message.

vsavkin committed:
feat(forms): add support for Validator

Currently, the only way for a directive to export a validator is by providing a function. This makes it ackward to write validators that depend on directive inputs. In addition to supporting functions as validators, classes implementing the Validator interface are supported too.

vsavkin committed:
cleanup(forms): marks abstract classes as abstract

vsavkin committed:
cleanup(forms): minor cleanup

mary-poppins commented:
Merging PR #4778 on behalf of @vsavkin to branch [presubmit-vsavkin-pr-4778](https://github.com/angular/angular/tree/presubmit-vsavkin-pr-4778).
